main/FatesHistoryInterfaceMod.F90

```diff
@@ -5570,39 +5647,107 @@ subroutine update_history_hifrq2(this,nc,nsites,sites,bc_in,bc_out,dt_tstep)
-            else
-               hio_c_stomata_si_age(io_si,ipa2) = 0._r8
-               hio_c_lblayer_si_age(io_si,ipa2) = 0._r8
-            end if
+  ! ===============================================================================================
 
-         end do
+  subroutine update_history_hifrq_subsite_ageclass(this,nsites,sites,dt_tstep)
 
-      enddo do_sites ! site loop
+    ! ---------------------------------------------------------------------------------
+    ! This subroutine is intended to update all history variables with upfreq ==
+    ! group_hifrq_complex (i.e., that have a dimension in addition to that for the site
+    ! level) that DO include age class. So, e.g., FATES_GPP_AP is updated here, but not
+    ! FATES_GPP or FATES_GPP_PF. This is an expensive call; the model will probably run
+    ! much faster if the user is not using any of these diagnostics.
+    ! ---------------------------------------------------------------------------------
 
-    end associate
+    !
+    ! Arguments
+    class(fates_history_interface_type)             :: this
+    integer                 , intent(in)            :: nsites
+    type(ed_site_type)      , intent(inout), target :: sites(nsites)
+    real(r8)                , intent(in)            :: dt_tstep
+
+    type(fates_cohort_type), pointer :: ccohort
+    type(fates_patch_type),  pointer :: cpatch
+    integer :: s, io_si
+    real(r8) :: site_canopy_area
+    real(r8) :: dt_tstep_inv          ! Time step in frequency units (/s)
+    real(r8) :: patch_canarea_div_site_canarea  ! Weighting based on patch canopy area relative to site canopy area
+    real(r8) :: cohort_n_div_site_area          ! Weighting based on cohort density relative to site area
+
+    associate( &
+       hio_c_lblayer_si_age => this%hvars(ih_c_lblayer_si_age)%r82d, &
+       hio_c_stomata_si_age => this%hvars(ih_c_stomata_si_age)%r82d, &
+       hio_gpp_si_age       => this%hvars(ih_gpp_si_age)%r82d, &
+       hio_npp_si_age       => this%hvars(ih_npp_si_age)%r82d &
+    )
+
+    dt_tstep_inv = 1.0_r8 / dt_tstep
+
+    do_sites: do s = 1,nsites
+
+       ! Get site-wide canopy area
+       site_canopy_area = 0._r8
+       cpatch => sites(s)%oldest_patch
+       do while(associated(cpatch))
+          site_canopy_area = site_canopy_area + cpatch%total_canopy_area
+          cpatch => cpatch%younger
+       end do
+
+       io_si  = sites(s)%h_gid
+
+       ! Get ageclass-stratified variables
+       cpatch => sites(s)%oldest_patch
+       do while(associated(cpatch))
+
+          ! Canopy resistance terms
+          if (site_canopy_area .gt. nearzero) then
+             patch_canarea_div_site_canarea = cpatch%total_canopy_area / site_canopy_area
+             hio_c_stomata_si_age(io_si,cpatch%age_class) = &
+                  hio_c_stomata_si_age(io_si,cpatch%age_class) + &
+                  cpatch%c_stomata * mol_per_umol &
+                  * patch_canarea_div_site_canarea
```

Comment:
Apologies if this was discussed elsewhere and I missed it.  The previous version of this calculation was dividing through the sum of the `total_canopy_area` just for the given age class: 
https://github.com/NGEET/fates/blob/1be3962933351b94fef134d607ac7d821d884f96/main/FatesHistoryInterfaceMod.F90#L5603-L5606
https://github.com/NGEET/fates/blob/1be3962933351b94fef134d607ac7d821d884f96/main/FatesHistoryInterfaceMod.F90#L5297-L5298

This change appears to introduce dividing the by the `site_canopy_area`.  Do I have that correct and is that the intent?